### PR TITLE
AEA-826 tac negotiation skill w/ contract

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR finished tac skills to work again w/ contract (erc1155). It also makes modifications on aea modules based on user feedback and to improve logging output.

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

#1626

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -455,6 +455,17 @@ def to_latest(self) -> "PublicId":
...
+        Check if a string is a public id.
+
+        :param public_id_string: the public id in string format.
+        :return: bool indicating validity
+        """
+        match = re.match(cls.PUBLIC_ID_REGEX, public_id_string)
+        return match is not None
+
     @classmethod
     def from_str(cls, public_id_string: str) -> "PublicId":
```
Comment (solarw) [inline aea/configurations/base.py:470]: why not to use `is_valid_str` in `from_str` method?

File: aea/crypto/ethereum.py [hunk 2]
```diff
@@ -506,6 +506,15 @@ def try_estimate_gas(self, tx: Dict[str, Any]) -> Dict[str, Any]:
             _default_logger.debug("Error when trying to estimate gas: {}".format(e))
         return tx
 
+    @classmethod
+    def is_valid_address(cls, address: Address) -> bool:
```
Comment (solarw) [inline aea/crypto/ethereum.py:510]: add validation to ledger methods to check the address on transaction construction or balance check?

File: aea/helpers/search/models.py [hunk 3]
```diff
@@ -66,6 +66,12 @@ def __eq__(self, other):
             return False  # pragma: nocover
         return self.latitude == other.latitude and self.longitude == other.longitude
 
+    def __str__(self):
```
Comment (solarw) [inline aea/helpers/search/models.py:69]: to add?
```
def __repr__ (self):
   return str(self)
```
some times useful when print  a set of objects

File: aea/mail/base.py [hunk 4]
```diff
@@ -340,6 +344,7 @@ def to(self) -> Address:
     @to.setter
     def to(self, to: Address) -> None:
         """Set address of receiver."""
+        enforce(isinstance(to, str), f"To must be string. Found '{type(to)}'")
```
Comment (solarw) [inline aea/mail/base.py:347]: may be `isinstance(to, Address)`? i know that is an alias for str

Comment (solarw) [reply aea/mail/base.py:347]: same thing in various places

**Review by solarw — APPROVED** [review]
LGTM. minor comments

File: aea/crypto/ethereum.py (hunk 2 above)
Comment (DavidMinarsch) [reply aea/crypto/ethereum.py:510]: will do in a separate pr

File: aea/helpers/search/models.py (hunk 3 above)
Comment (DavidMinarsch) [reply aea/helpers/search/models.py:69]: will do in separate pr

File: aea/configurations/base.py (hunk 1 above)
Comment (DavidMinarsch) [reply aea/configurations/base.py:470]: is_valid_str only returns bool, from_str needs match

File: aea/mail/base.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/mail/base.py:347]: yes that could one done eventually when we consistently apply Address

**Pull request merged**